### PR TITLE
unbreak release build: no side effects in JLM_ASSERT

### PR DESCRIPTION
**Pull request by caleridas**
Operations that have side effect should not go into JLM_ASSERT.
Otherwise, they will not be executed in release builds.

### CODE REVIEW

File: jlm/rvsdg/node.cpp [hunk 1]
```diff
@@ -113,7 +113,7 @@ output::remove_user(jlm::rvsdg::input * user)
   {
     if (!node->has_users())
     {
-      JLM_ASSERT(region()->AddBottomNode(*node));
+      region()->AddBottomNode(*node);
```
Comment (phate) [inline jlm/rvsdg/node.cpp:116]: Could we do something like this for all of them instead:

```
auto wasAdded = region()->AddBottomNode(*node);
JLM_ASSERT(wasAdded)
```

**Review by phate — CHANGES_REQUESTED** [review]
Uff that is a big one. It would have been caught if issue #579 would have been done already.

File: jlm/rvsdg/node.cpp (hunk 1 above)
Comment (caleridas) [reply jlm/rvsdg/node.cpp:116]: Doesn't this risk unused variable warnings in release builds instead?

Comment (haved) [reply jlm/rvsdg/node.cpp:116]: I don't think so, as the definition of `JLM_ASSERT` is
```c++
#define JLM_ASSERT(x) \
  do                  \
  {                   \
  } while (0 && (x))
```

Which means the expression in still "used", without being evaluated. Unless the compiler is eliminating dead code before checking, it should not give a warning.

I tested the following in g++ 14.2.1 and clang++ 18.1.8, with `-Wall -Wextra`, and they only complain about `z`.
```c++
int main() {
    int x = 5;
    int y = 7;
    int z = 10;

    do {
    } while(0 & (x));

    return y;
}
```

Comment (caleridas) [reply jlm/rvsdg/node.cpp]: oh right -- changed it accordingly

**Review by phate — APPROVED** [review]

**Pull request merged**